### PR TITLE
HW3 Task 1

### PR DESCRIPTION
**Pull request by ValeriL**
There are #noqa due to Cognitive complexity errors (its <10 but >5). I tried to fix it but failed(

### CODE REVIEW

File: Homework/homework3/task_1.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    def decorator(func: Callable) -> Callable:  # noqa
+        cache_memory = []
+
+        def caching(*args, **kwargs):  # noqa
+            cur_input = (args, kwargs)
+            for stored_args, result, count_times in cache_memory:
+                if stored_args == cur_input:
+                    cache_memory.remove((stored_args, result, count_times))
+                    if count_times < times:
+                        cache_memory.append((stored_args, result, count_times + 1))
```
Comment (Cjkjvfnby) [inline Homework/homework3/task_1.py:14]: It is more easy to decrement, start form times, and end at 0.

File: Homework/homework3/test_task1.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+from unittest.mock import Mock
+
+from homework3.task_1 import cache
+
+import pytest
+
+
+@pytest.mark.parametrize(
+    "times,calls,expected_result", [(1, 4, 2), (2, 4, 2), (3, 4, 1), (3, 5, 2)]
```
Comment (Cjkjvfnby) [inline Homework/homework3/test_task1.py:9]: What the difference between tests?  Does the next one bring additional value?

File: Homework/homework3/test_task1.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+)
+def test_cache(times: int, calls: int, expected_result: int):
+
+    mock = Mock()
+    cache_call = cache(times)(mock)
+
+    for _ in range(0, calls):
```
Comment (Cjkjvfnby) [inline Homework/homework3/test_task1.py:16]: Do not use a cycle here.  Just call functions manually and put an assert in the middle.

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]

File: Homework/homework3/test_task1.py (hunk 2 above)
Comment (ValeriL) [reply Homework/homework3/test_task1.py:9]: Just to ensure that its work well with different values of times and different number of calls of function (for not calling function manually).

**Review by Cjkjvfnby — APPROVED** [review]

**Pull request merged**